### PR TITLE
[Dataflow Streaming] Use separate heartbeat streams based on job settings

### PR DESCRIPTION
**Pull request by arunpandianp**
The logic is guarded behind an experiment `streaming_engine_use_job_settings_for_heartbeat_pool`. The experiment will be removed once backend changes are rollback safe.

### CODE REVIEW

**Comment by arunpandianp** [comment]
R: @scwhittle

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 1]
```diff
@@ -253,12 +255,29 @@ private StreamingDataflowWorker(
...
+                ? StreamPoolHeartbeatSender.Create(
+                    options.getUseSeparateWindmillHeartbeatStreams()
+                        ? separateHeartbeatPool(windmillServer)
+                        : getDataStreamPool)
+                : StreamPoolHeartbeatSender.Create(
+                    separateHeartbeatPool(windmillServer),
+                    getDataStreamPool,
+                    configFetcher.getGlobalConfigHandle());
+      } else {
+        heartbeatSender =
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:274]: think you can share some of the cases
```
if (!hasExperiment || options.getUseSeparateWindmillHeartbeatStreams() != null) {
              StreamPoolHeartbeatSender.Create(
                Boolean.TRUE.equals(options.getUseSeparateWindmillHeartbeatStreams())
                    ? separateHeartbeatPool(windmillServer)
                    : getDataStreamPool);
} else {
StreamPoolHeartbeatSender.Create(
                    separateHeartbeatPool(windmillServer),
                    getDataStreamPool,
                    configFetcher.getGlobalConfigHandle());
}
```

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 2]
```diff
@@ -253,12 +255,29 @@ private StreamingDataflowWorker(
...
       getDataClient = new StreamPoolGetDataClient(getDataMetricTracker, getDataStreamPool);
-      heartbeatSender =
-          new StreamPoolHeartbeatSender(
-              options.getUseSeparateWindmillHeartbeatStreams()
-                  ? WindmillStreamPool.create(
-                      1, GET_DATA_STREAM_TIMEOUT, windmillServer::getDataStream)
-                  : getDataStreamPool);
+      // Experiment gates the logic till backend changes are rollback safe
+      if (DataflowRunner.hasExperiment(
+          options, STREAMING_ENGINE_USE_JOB_SETTINGS_FOR_HEARTBEAT_POOL)) {
```
[the remark this answers is not in the document]
Comment (scwhittle) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:260]: Discussed, leaving as is and will remove once safe. Defaulting to on will let us cleanup eventually and have released sdks use preferred behavior.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 1 above)
Comment (arunpandianp) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:262]: Done.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/work/refresh/StreamPoolHeartbeatSender.java [hunk 3]
```diff
@@ -27,19 +30,52 @@
...
+                    ? heartbeatStreamPool
+                    : getDataPool));
+    return heartbeatSender;
   }
 
   @Override
   public void sendHeartbeats(Heartbeats heartbeats) {
     try (CloseableStream<WindmillStream.GetDataStream> closeableStream =
-        heartbeatStreamPool.getCloseableStream()) {
+        heartbeatStreamPool.get().getCloseableStream()) {
```
[the remark this answers is not in the document]
Comment (arunpandianp) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/work/refresh/StreamPoolHeartbeatSender.java:79]: The settings callback has a reference to both the pools and it switches between them whenever the setting changes.

[the remark this answers is not in the document]
Comment (arunpandianp) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/work/refresh/StreamPoolHeartbeatSender.java:79]: renamed variables to make it more clear.

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/work/refresh/StreamPoolHeartbeatSenderTest.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            WindmillStreamPool.create(
+                1, Duration.standardSeconds(10), getDataServer::getDataStream),
+            configHandle);
+    Heartbeats.Builder heartbeatsBuilder = Heartbeats.builder();
+    heartbeatsBuilder
+        .heartbeatRequestsBuilder()
+        .put("key", HeartbeatRequest.newBuilder().setWorkToken(123).build());
+    heartbeatSender.sendHeartbeats(heartbeatsBuilder.build());
+    assertEquals(0, dedicatedServer.getGetDataRequests().size());
+    assertEquals(1, getDataServer.getGetDataRequests().size());
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/work/refresh/StreamPoolHeartbeatSenderTest.java:111]: then set the config to different setting and verify it goes to the other server for next heartbeat sent on the same heartbeatSender object?

**Review by scwhittle — CHANGES_REQUESTED** [review]

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/work/refresh/StreamPoolHeartbeatSenderTest.java (hunk 4 above)
Comment (arunpandianp) [reply runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/work/refresh/StreamPoolHeartbeatSenderTest.java:119]: Done.

**Comment by arunpandianp** [comment]
Run Java Precommit

**Comment by scwhittle** [comment]
precommit is unrelated pubsub auth issue.

**Review by scwhittle — APPROVED** [review]

**Pull request merged**